src/runes/runestone.rs

```diff
@@ -0,0 +1,501 @@
+use super::*;
+
+#[derive(Default, Serialize, Debug, PartialEq)]
+pub struct Runestone {
+  pub edicts: Vec<Edict>,
+  pub etching: Option<Etching>,
+}
+
+impl Runestone {
+  pub fn from_transaction(transaction: &Transaction) -> Option<Self> {
+    Self::decipher(transaction).ok().flatten()
+  }
+
+  fn decipher(transaction: &Transaction) -> Result<Option<Self>> {
+    let Some(payload) = Runestone::payload(transaction)? else {
+      return Ok(None);
+    };
+
+    let mut integers = Vec::new();
+    let mut i = 0;
+
+    while i < payload.len() {
+      let (integer, length) = varint::decode(&payload[i..])?;
+      integers.push(integer);
+      i += length;
+    }
+
+    let mut edicts = Vec::new();
+    let mut etching = None;
+
+    for chunk in integers.chunks(3) {
+      match *chunk {
```

Comment:
maybe think about reifying the `[rune]`, `[rune, decimals]`, `[id, amount, output]` variants here into an enum?

Reply:
I actually think this is probably okay, just because it makes it clear how it works, i.e., extra values are treated as rune, or rune and decimals, and adding an enum might make it less clear.